scripts/publish.py

```diff
@@ -0,0 +1,123 @@
+#! /bin/python3
+
+# This is a script for publishing the wasmer crates to crates.io.
+# It should be run in the root of wasmer like `python3 scripts/publish.py`.
+
+# install dependencies:
+# pip3 install toposort
+
+import argparse
+import os
+import re
+import subprocess
+import time
+
+from typing import Optional
+from toposort import toposort_flatten
+
+
+# TODO find this automatically
+target_version = "1.0.0-alpha01.1"
+
+# TODO generate this by parsing toml files
+dep_graph = {
+    "wasmer-types": set([]),
+    "wasmer-vm": set(["wasmer-types"]),
+    "wasmer-compiler": set(["wasmer-vm", "wasmer-types"]),
+    "wasmer-object": set(["wasmer-types", "wasmer-compiler"]),
+    "wasmer-engine": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-singlepass": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-cranelift": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-llvm": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-engine-jit": set(["wasmer-types", "wasmer-vm", "wasmer-compiler", "wasmer-engine"]),
+    "wasmer-engine-native": set(["wasmer-types", "wasmer-vm", "wasmer-compiler", "wasmer-engine",
+                                 "wasmer-object"]),
+    "wasmer": set(["wasmer-vm", "wasmer-compiler-singlepass", "wasmer-compiler-cranelift",
+                   "wasmer-compiler-llvm", "wasmer-compiler", "wasmer-engine", "wasmer-engine-jit",
+                   "wasmer-engine-native", "wasmer-types"]),
+    "wasmer-cache": set(["wasmer"]),
+    "wasmer-wasi": set(["wasmer"]),
+    "wasmer-wasi-experimental-io-devices": set(["wasmer-wasi"]),
+    "wasmer-emscripten": set(["wasmer"]),
+}
+
+# where each crate is located in `lib`
+# this can also be generated from the toml files
+location = {
+    "wasmer-types": "wasmer-types",
+    "wasmer-vm": "vm",
+    "wasmer-compiler": "compiler",
+    "wasmer-object": "object",
+    "wasmer-engine": "engine",
+    "wasmer-compiler-singlepass": "compiler-singlepass",
+    "wasmer-compiler-cranelift": "compiler-cranelift",
+    "wasmer-compiler-llvm": "compiler-llvm",
+    "wasmer-engine": "engine",
+    "wasmer-engine-jit": "engine-jit",
+    "wasmer-engine-native": "engine-native",
+    "wasmer-cache": "cache",
+    "wasmer": "api",
+    "wasmer-wasi": "wasi",
+    "wasmer-emscripten": "emscripten",
+    "wasmer-wasi-experimental-io-devices": "wasi-experimental-io-devices",
+}
+
+dry_run = False
```

Comment:
Default to True, rename `dry_run` arg to `nodry_run`?

Reply:
So I based the command on `cargo publish` which also has a `--dry-run` flag and defaults to publishing, which is why that made sense to me

Reply:
I would still prefer to default to a dry run here. Some tools like python linters work by running the code (hence the `if name == "__main__":` at the end) or someone may be editing the code or refactoring it into a larger library and we want to avoid accidents. I agree that having a `dry_run` is the normal approach and our reasons don't apply to `cargo publish` itself, but I'd still rather this program and tools like it operate in "safe by default" mode with an explicit decision to turn on the business end. To me it's less about this script and more about a general rule. I could live with us keeping `--dry_run` here if you feel strongly about it.

Reply:
Hmm, interesting point, the main reason why I think not defaulting `dry-run` makes sense is that `dry-run` is only useful when editing the script which I perceive to be a less common operation than running it.

I'm seeing this script more as a quick hack like the `scripts/update-version.sh` I made than as a robust tool: it's a simple piece of code that does the right thing enough that it's a useful way to save some time.

I don't feel that strongly though and we can make it opt-out of test mode

Reply:
> I'm seeing this script more as a quick hack like the scripts/update-version.sh I made than as a robust tool: it's a simple piece of code that does the right thing enough that it's a useful way to save some time.

This is only temporary unless it works. :)

Up to you whether to change it.